src/toil/batchSystems/mesos/__init__.py

```diff
@@ -27,22 +31,110 @@
     'cores'))
 
 
-class ResourceRequirement( namedtuple('_ResourceRequirement', (
-    # Number of bytes (!) needed for a task
-    'memory',
-    # Number of CPU cores needed for a task
-    'cores',
-    # Number of bytes (!) needed for task on disk
-    'disk',
-    # True, if job can be run on a preemptable node, False otherwise
-    'preemptable'))):
+class JobQueue(object):
+
+    def __init__(self):
+        # mapping of jobTypes to queues of jobs of that type
+        self.queues = {}
+        # list of jobTypes in decreasing resource expense
+        self.sortedTypes = []
+
+    def insertJob(self, job, jobType):
+        # logn insertion
+        if jobType not in self.queues:
+            index = bisect(self.sortedTypes, jobType)
+            self.sortedTypes.insert(index, jobType)
+            self.queues[jobType] = Queue()
+
+        self.queues[jobType].put(job)
+
+    def sorted(self):
+        return self.sortedTypes
+
+    def jobIDs(self):
+        # O(N)
+        return [job.jobID for queue in self.queues.values() for job in list(queue.queue)]
+
+    def nextJobOfType(self, jobType):
+        # constant lookup, pop off first element
+        # also remove from taskMap
+        job = self.queues[jobType].get(block=False)
```

Comment:
Unless I missed something, this will end up leaking jobTypes/Queues (keeping them around in `sortedTypes` and `queues`) even after all the jobs of a given type have been run. The leaked memory isn't really an issue, but the leaked entries in `sortedTypes` will cause a slowdown. The scheduler will have to iterate over each empty jobType. If there are a ton of different jobTypes, most of which only have one or two jobs associated with them, the leader will get slower and slower as time goes on.

Fixing this would be easy--just remove the jobType once its queue runs out--but would probably require a lock on the list & hash to avoid losing jobs.